### PR TITLE
Ported JsonWriterPerfTests to BenchmarkDotNet

### PR DESCRIPTION
**Pull request by ahsonkhan**
``` ini

BenchmarkDotNet=v0.10.14.525-nightly, OS=Windows 10.0.16299.371 (1709/FallCreatorsUpdate/Redstone3)
Intel Xeon CPU E5-1620 v2 3.70GHz, 1 CPU, 8 logical and 4 physical cores
Frequency=3604597 Hz, Resolution=277.4235 ns, Timer=TSC
.NET Core SDK=2.1.300-rc1-008662
  [Host]     : .NET Core 2.2.0-preview1-26424-02 (CoreCLR 4.6.26423.05, CoreFX 4.6.26423.04), 64bit RyuJIT
  DefaultJob : .NET Core 2.2.0-preview1-26424-02 (CoreCLR 4.6.26423.05, CoreFX 4.6.26423.04), 64bit RyuJIT


```
|                         Method |         Target | Formatted |        Mean |        Error |        StdDev |  Gen 0 | Allocated |
|------------------------------- |--------------- |---------- |------------:|-------------:|--------------:|-------:|----------:|
|      **WriterSystemTextJsonBasic** | **InvariantUtf16** |     **False** | **72,467.4 ns** | **1,360.429 ns** | **1,397.0607 ns** |      **-** |       **0 B** |
|          WriterNewtonsoftBasic | InvariantUtf16 |     False | 18,095.1 ns |   403.423 ns |   414.2858 ns | 0.0610 |     416 B |
| WriterSystemTextJsonHelloWorld | InvariantUtf16 |     False |    548.0 ns |     3.680 ns |     3.0733 ns |      - |       0 B |
|     WriterNewtonsoftHelloWorld | InvariantUtf16 |     False |    332.7 ns |     5.978 ns |     4.3226 ns | 0.0486 |     256 B |
|      **WriterSystemTextJsonBasic** | **InvariantUtf16** |      **True** | **97,282.2 ns** | **1,771.764 ns** | **1,570.6226 ns** |      **-** |       **0 B** |
|          WriterNewtonsoftBasic | InvariantUtf16 |      True | 30,431.1 ns |   465.610 ns |   388.8057 ns | 0.0916 |     584 B |
| WriterSystemTextJsonHelloWorld | InvariantUtf16 |      True |    749.0 ns |    12.867 ns |    12.0354 ns |      - |       0 B |
|     WriterNewtonsoftHelloWorld | InvariantUtf16 |      True |    386.5 ns |     3.193 ns |     2.4933 ns | 0.0806 |     424 B |
|      **WriterSystemTextJsonBasic** |  **InvariantUtf8** |     **False** | **70,201.9 ns** | **1,387.989 ns** | **1,298.3261 ns** |      **-** |       **0 B** |
|          WriterNewtonsoftBasic |  InvariantUtf8 |     False | 18,153.2 ns |   261.378 ns |   218.2622 ns | 0.0610 |     416 B |
| WriterSystemTextJsonHelloWorld |  InvariantUtf8 |     False |    567.2 ns |     1.149 ns |     0.8309 ns |      - |       0 B |
|     WriterNewtonsoftHelloWorld |  InvariantUtf8 |     False |    290.7 ns |     3.006 ns |     2.6650 ns | 0.0486 |     256 B |
|      **WriterSystemTextJsonBasic** |  **InvariantUtf8** |      **True** | **96,535.6 ns** | **1,713.813 ns** | **1,603.1017 ns** |      **-** |       **0 B** |
|          WriterNewtonsoftBasic |  InvariantUtf8 |      True | 31,116.4 ns |   120.934 ns |    87.4434 ns | 0.0610 |     584 B |
| WriterSystemTextJsonHelloWorld |  InvariantUtf8 |      True |    741.8 ns |    14.649 ns |    14.3872 ns |      - |       0 B |
|     WriterNewtonsoftHelloWorld |  InvariantUtf8 |      True |    379.1 ns |     7.390 ns |     7.9071 ns | 0.0806 |     424 B |

**The following PR regressed the performance of the JsonWriter 2x.**
Change IOutput to merge methods and return Memory - https://github.com/dotnet/corefxlab/pull/2047

cc @pakrym, @davidfowl 

https://github.com/dotnet/corefxlab/commit/566d82abe5845fef0c421acd08b0b8d0bd04c745 - Jan 16, 2018 ([+194 from today](https://github.com/dotnet/corefxlab/commits/master?after=79b72129cf1184ed3945e717fed3003889204a38+194))
- WriterNewtonsoftBasic(formatted: True, encoderTarget: InvariantUtf8) 52.933
- WriterSystemTextJsonBasic(formatted: True, encoderTarget: InvariantUtf8) 49.065

https://github.com/dotnet/corefxlab/commit/4bf5f6a527c109c888a78ff8b2593ee631f33126 - Jan 17, 2018 ([+193 from today](https://github.com/dotnet/corefxlab/commits/master?after=79b72129cf1184ed3945e717fed3003889204a38+193))
- WriterNewtonsoftBasic(formatted: True, encoderTarget: InvariantUtf8) 52.230
- WriterSystemTextJsonBasic(formatted: True, encoderTarget: InvariantUtf8) 97.776

### CODE REVIEW

File: tests/Benchmarks/System.Text.Json/JsonWriterPerf.cs [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+        [ArgumentsSource(nameof(GetWriterSystemTextJsonParameters))]
+        public void WriterSlowSystemTextJsonBasic(bool formatted, ArrayFormatter formatter)
+        {
+            formatter.Clear();
+            WriterSystemTextJsonBasic(formatted, formatter);
+        }
+
+        // Cannot use this since ArrayFormatter is not a compile time constant
+        // System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.MissingMethodException: No parameterless constructor defined for this object.
+        public IEnumerable<object[]> GetWriterSystemTextJsonParameters()
```
Comment (ahsonkhan) [inline tests/Benchmarks/System.Text.Json/JsonWriterPerf.cs]: @adamsitnik, how can I create arguments that are not compile-time constants? I tried to implement IParam instead but was running into some issues. Any suggestions to get me on the right track?

**Comment by ahsonkhan** [comment]
With the change from https://github.com/dotnet/corefxlab/pull/2255, here are the results.
``` ini

BenchmarkDotNet=v0.10.14.525-nightly, OS=Windows 10.0.16299.371 (1709/FallCreatorsUpdate/Redstone3)
Intel Xeon CPU E5-1620 v2 3.70GHz, 1 CPU, 8 logical and 4 physical cores
Frequency=3604597 Hz, Resolution=277.4235 ns, Timer=TSC
.NET Core SDK=2.1.300-rc1-008662
  [Host]     : .NET Core 2.2.0-preview1-26424-02 (CoreCLR 4.6.26423.05, CoreFX 4.6.26423.04), 64bit RyuJIT
  DefaultJob : .NET Core 2.2.0-preview1-26424-02 (CoreCLR 4.6.26423.05, CoreFX 4.6.26423.04), 64bit RyuJIT


```
|                         Method |         Target | Formatted |        Mean |         Error |        StdDev |  Gen 0 | Allocated |
|------------------------------- |--------------- |---------- |------------:|--------------:|--------------:|-------:|----------:|
|      **WriterSystemTextJsonBasic** | **InvariantUtf16** |     **False** | **53,124.9 ns** |   **305.2463 ns** |   **254.8944 ns** |      **-** |       **0 B** |
|          WriterNewtonsoftBasic | InvariantUtf16 |     False | 18,012.4 ns |   282.4668 ns |   220.5315 ns | 0.0610 |     416 B |
| WriterSystemTextJsonHelloWorld | InvariantUtf16 |     False |    409.1 ns |     0.7588 ns |     0.6336 ns |      - |       0 B |
|     WriterNewtonsoftHelloWorld | InvariantUtf16 |     False |    284.9 ns |     2.4192 ns |     2.1445 ns | 0.0486 |     256 B |
|      **WriterSystemTextJsonBasic** | **InvariantUtf16** |      **True** | **69,077.0 ns** |   **387.1479 ns** |   **343.1965 ns** |      **-** |       **0 B** |
|          WriterNewtonsoftBasic | InvariantUtf16 |      True | 30,415.9 ns |    78.3646 ns |    61.1819 ns | 0.0916 |     584 B |
| WriterSystemTextJsonHelloWorld | InvariantUtf16 |      True |    549.1 ns |     1.7536 ns |     1.4643 ns |      - |       0 B |
|     WriterNewtonsoftHelloWorld | InvariantUtf16 |      True |    388.6 ns |     7.6148 ns |     8.4638 ns | 0.0806 |     424 B |
|      **WriterSystemTextJsonBasic** |  **InvariantUtf8** |     **False** | **52,667.2 ns** |   **281.3012 ns** |   **219.6215 ns** |      **-** |       **0 B** |
|          WriterNewtonsoftBasic |  InvariantUtf8 |     False | 18,514.1 ns |   336.5748 ns |   298.3648 ns | 0.0610 |     416 B |
| WriterSystemTextJsonHelloWorld |  InvariantUtf8 |     False |    431.5 ns |     2.2382 ns |     1.8690 ns |      - |       0 B |
|     WriterNewtonsoftHelloWorld |  InvariantUtf8 |     False |    294.3 ns |     5.5165 ns |     5.1601 ns | 0.0486 |     256 B |
|      **WriterSystemTextJsonBasic** |  **InvariantUtf8** |      **True** | **68,444.1 ns** | **1,290.7185 ns** | **1,434.6300 ns** |      **-** |       **0 B** |
|          WriterNewtonsoftBasic |  InvariantUtf8 |      True | 31,128.0 ns |   269.8018 ns |   225.2967 ns | 0.0610 |     584 B |
| WriterSystemTextJsonHelloWorld |  InvariantUtf8 |      True |    560.4 ns |     1.0347 ns |     0.7481 ns |      - |       0 B |
|     WriterNewtonsoftHelloWorld |  InvariantUtf8 |      True |    387.7 ns |     0.8679 ns |     0.7247 ns | 0.0806 |     424 B |

**Comment by KrzysztofCwalina** [comment]
BTW, it would be good to add "true UTF8" tests, i.e. tests writing/reading UTF8 literals. 

cc: @GrabYourPitchforks

File: tests/Benchmarks/System.Text.Json/JsonWriterPerf.cs (hunk 1 above)
Comment (adamsitnik) [reply tests/Benchmarks/System.Text.Json/JsonWriterPerf.cs]: @ahsonkhan I must admit that this is one of the BDN pitfalls (the compilation time requirement for arguments). What you did (using primitive field params) is the best way to do it as of today. Implementing `IParam` requires more work.

I will try to think about more user friendly solution, but as of today I don't have any better idea.

**Pull request merged**